upstairs/src/downstairs.rs

```diff
@@ -3673,6 +3821,589 @@ impl Downstairs {
+        use omicron_uuid_kinds::GenericUuid;
+        use omicron_uuid_kinds::TypedUuid;
+        use omicron_uuid_kinds::UpstairsKind;
+
+        let upstairs_id: TypedUuid<UpstairsKind> =
+            TypedUuid::from_untyped_uuid(self.cfg.upstairs_id);
+
+        let Some(downstairs_id) = self.clients[client_id].id() else {
+            return;
+        };
+        let downstairs_id: TypedUuid<DownstairsKind> =
+            TypedUuid::from_untyped_uuid(downstairs_id);
+
+        let now = Utc::now();
+        let log = self
+            .log
+            .new(o!("downstairs_id" => downstairs_id.to_string()));
+
+        let reason = match reason {
+            ClientRunResult::ConnectionTimeout => {
+                DownstairsClientStoppedReason::ConnectionTimeout
+            }
+            ClientRunResult::ConnectionFailed(_) => {
+                DownstairsClientStoppedReason::ConnectionFailed
+            }
+            ClientRunResult::Timeout => DownstairsClientStoppedReason::Timeout,
+            ClientRunResult::WriteFailed(_) => {
+                DownstairsClientStoppedReason::WriteFailed
+            }
+            ClientRunResult::ReadFailed(_) => {
+                DownstairsClientStoppedReason::ReadFailed
+            }
+            ClientRunResult::RequestedStop(_) => {
+                DownstairsClientStoppedReason::RequestedStop
+            }
+            ClientRunResult::Finished => {
+                DownstairsClientStoppedReason::Finished
+            }
+            ClientRunResult::QueueClosed => {
+                DownstairsClientStoppedReason::QueueClosed
+            }
+            ClientRunResult::ReceiveTaskCancelled => {
+                DownstairsClientStoppedReason::ReceiveTaskCancelled
+            }
+        };
+
+        // Spawn a task so we don't block the main loop talking to
+        // Nexus.
+        let target_addrs = self.get_target_addrs();
+        tokio::spawn(async move {
+            let Some(nexus_client) =
+                get_nexus_client(&log, &target_addrs).await
+            else {
+                // Exit if no Nexus client returned from DNS - our notification
+                // is best effort.
+                return;
+            };
+
+            match omicron_common::retry_until_known_result(&log, || async {
+                nexus_client
+                    .cpapi_downstairs_client_stopped(
+                        &upstairs_id,
+                        &downstairs_id,
+                        &DownstairsClientStopped { time: now, reason },
+                    )
+                    .await
+            })
+            .await
+            {
+                Ok(_) => {
+                    info!(log, "notified Nexus of client stopped");
+                }
+
+                Err(e) => {
+                    error!(
+                        log,
+                        "failed to notify Nexus of client stopped: {e}"
+                    );
+                }
+            }
+        });
```

Comment:
I wish there was a way to reduce copy-pasta in this pattern, but got into a terrible mess of futures and lifetimes when I tried to write a generic `notify_nexus(..)`

Reply:
Yeah, it's unfortunate. I figured such a clean up could occur somewhere down the road if it was possible, but wanted to get this PR out.